pyuvdata/uvdata.py

```diff
@@ -4628,3 +4654,497 @@ def inflate_by_redundancy(self, tol=1.0, blt_order='time', blt_minor_order=None)
+                                  "gaps.".format(bl=self.baseline_to_antnums(bl)))
+
+        temp_Nblts = n_new_samples
+
+        input_phase_type = self.phase_type
+        if input_phase_type == "drift":
+            if allow_drift:
+                print('Data are in drift mode and allow_drift is True, so '
+                      'resampling will be done without phasing.')
+            else:
+                # phase to RA/dec of zenith
+                print('Data are in drift mode, phasing before resampling.')
+                phase_time = Time(self.time_array[0], format='jd')
+                self.phase_to_time(phase_time)
+
+        # make temporary arrays
+        temp_baseline = np.zeros((temp_Nblts,), dtype=np.int)
+        temp_time = np.zeros((temp_Nblts,))
+        temp_uvw = np.zeros((temp_Nblts, 3))
+        temp_int_time = np.zeros((temp_Nblts,))
+        if not self.metadata_only:
+            temp_data = np.zeros((temp_Nblts, self.Nspws, self.Nfreqs, self.Npols),
+                                 dtype=self.data_array.dtype)
+            temp_flag = np.zeros((temp_Nblts, self.Nspws, self.Nfreqs, self.Npols),
+                                 dtype=self.flag_array.dtype)
+            temp_nsample = np.zeros((temp_Nblts, self.Nspws, self.Nfreqs, self.Npols),
+                                    dtype=self.nsample_array.dtype)
+
+        temp_idx = 0
+        for bl in bls_to_downsample:
+            bl_inds = np.nonzero(self.baseline_array == bl)[0]
+            running_int_time = 0.0
+            summing_idx = 0
+            n_sum = 0
+            for itime, int_time in enumerate(self.integration_time[bl_inds]):
+                running_int_time += int_time
+                n_sum += 1
+                over_min_int_time = (running_int_time > min_int_time
+                                     or np.isclose(running_int_time, min_int_time,
+                                                   rtol=self._integration_time.tols[0],
+                                                   atol=self._integration_time.tols[1]))
+                last_sample = (itime == len(bl_inds) - 1)
+                # We sum up all the samples found so far if we're over the target minimum
+                # time, or we've hit the end of the time samples for this baseline.
+                if over_min_int_time or last_sample:
+                    if last_sample and not (over_min_int_time or keep_ragged):
+                        # don't do anything -- implicitly drop these integrations
+                        continue
+                    # sum together that number of samples
+                    temp_baseline[temp_idx] = bl
+                    # this might be wrong if some of the constituent times are *totally* flagged
+                    averaging_idx = bl_inds[summing_idx:summing_idx + n_sum]
+                    temp_time[temp_idx] = np.average(self.time_array[averaging_idx])
```

Comment:
I think there's an implicit assumption here (and line 5006) that the integration times are uniform within this baseline. Here I think it can be fixed with something like:
```
temp_time[temp_idx] = (np.sum(self.time_array[averaging_idx] * self.integration_integration_time[averaging_idx])
                       / np.sum(self.integration_integration_time[averaging_idx])
```
And a similar weighted mean would make sense below.

Reply:
Good point! I've updated it as indicated here and in the actual data averaging below (which was slightly trickier to get right because of the mis-match in array dimensions between the data and `integration_time`).